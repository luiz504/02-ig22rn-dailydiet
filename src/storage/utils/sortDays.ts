export const sortDays = (array: string[]) => {
  return array.sort((a, b) => {
    return new Date(b).getTime() - new Date(a).getTime()
  })
}
