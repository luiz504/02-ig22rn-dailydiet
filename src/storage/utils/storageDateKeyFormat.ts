import { format } from 'date-fns'

export function storageDateKeyFormat(date: Date): string {
  return format(date, 'yyyy/MM/dd')
}
