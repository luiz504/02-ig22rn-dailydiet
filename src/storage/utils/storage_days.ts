import AsyncStorage from '@react-native-async-storage/async-storage'
import { DAY_COLLECTION } from '../commons'

export async function getStoredDays() {
  return AsyncStorage.getItem(DAY_COLLECTION).then((storedData) =>
    storedData ? (JSON.parse(storedData) as string[]) : null,
  )
}

export async function setStoredDays(dayCollection: string[]) {
  return AsyncStorage.setItem(DAY_COLLECTION, JSON.stringify(dayCollection))
}
