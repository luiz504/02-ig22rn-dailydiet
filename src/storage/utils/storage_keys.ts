import { format, parse } from 'date-fns'

const FORMAT = 'yyyy/MM/dd'
export function storageDateKeyFormat(date: Date): string {
  return format(date, FORMAT)
}

export function storageKeyToDateParse(key: string) {
  return parse(key, FORMAT, new Date())
}
