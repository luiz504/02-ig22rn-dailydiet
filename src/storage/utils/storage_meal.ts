import AsyncStorage from '@react-native-async-storage/async-storage'
import { MEAL_COLLECTION } from '../commons'
import { Meal } from '~/models/Meal'

export type MealGroupByDay = {
  day: string
  meals: Meal[]
}

export async function getStoredMealsByDay(key: string) {
  return AsyncStorage.getItem(`${MEAL_COLLECTION}-${key}`).then((storedData) =>
    storedData ? (JSON.parse(storedData) as MealGroupByDay) : null,
  )
}

export async function setStoredMeals(key: string, meals: Meal[]) {
  const sortedMealsByHour = meals.sort(
    (a, b) => new Date(b.date).getTime() - new Date(a.date).getTime(),
  )
  const mealsCollection: MealGroupByDay = { day: key, meals: sortedMealsByHour }

  return AsyncStorage.setItem(
    `${MEAL_COLLECTION}-${key}`,
    JSON.stringify(mealsCollection),
  )
}

export async function getStoredMealsByDays(keys: string[]) {
  const keysCombined = keys.map((k) => `${MEAL_COLLECTION}-${k}`)
  const entries = await AsyncStorage.multiGet(keysCombined)
  return entries
    .map(([, value]) => (value ? JSON.parse(value) : null))
    .filter((value) => value) as MealGroupByDay[]
}

export async function deleteStoredMealCollection(key: string) {
  return AsyncStorage.removeItem(`${MEAL_COLLECTION}-${key}`)
}
