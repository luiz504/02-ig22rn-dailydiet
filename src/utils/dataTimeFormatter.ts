import * as Localization from 'expo-localization'

const userConfig = {
  uses24hoursClock:
    Localization?.getCalendars?.()?.[0]?.uses24hourClock ?? true,
  languageTag: Localization?.getLocales?.()?.[0]?.languageTag ?? 'en-US',
}

function formatDate(date: Date) {
  return new Intl.DateTimeFormat(userConfig.languageTag, {
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
  }).format(date)
}

function formatTime(time: Date) {
  return new Intl.DateTimeFormat(userConfig.languageTag, {
    hour: '2-digit',
    minute: '2-digit',
    hour12: !userConfig.uses24hoursClock,
  }).format(time)
}

export { userConfig, formatDate, formatTime }
