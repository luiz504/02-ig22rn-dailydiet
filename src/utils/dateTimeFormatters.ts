import * as Localization from 'expo-localization'

const getFirstUserConfig = () => ({
  uses24hoursClock: Localization.getCalendars()[0].uses24hourClock,
  languageTag: Localization.getLocales()[0].languageTag,
})

function formatDate(date: Date) {
  const { languageTag } = getFirstUserConfig()

  return new Intl.DateTimeFormat(languageTag, {
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
  }).format(date)
}

function formatTime(time: Date) {
  const { languageTag, uses24hoursClock } = getFirstUserConfig()

  return new Intl.DateTimeFormat(languageTag, {
    hour: '2-digit',
    minute: '2-digit',
    hour12: !uses24hoursClock,
  }).format(time)
}

function formatDateAndTime(date: Date | string) {
  const { languageTag, uses24hoursClock } = getFirstUserConfig()

  const formattedFullDate = new Date(date).toLocaleDateString(languageTag, {
    year: 'numeric',
    month: '2-digit',
    day: '2-digit',
    hour: '2-digit',
    minute: '2-digit',
    hour12: !uses24hoursClock,
  })

  if (languageTag.startsWith('en')) {
    return formattedFullDate.replace(',', ' at')
  }

  if (languageTag.startsWith('pt')) {
    return formattedFullDate.replace(',', ' Ã s')
  }

  return formattedFullDate
}

export { getFirstUserConfig, formatDate, formatTime, formatDateAndTime }
