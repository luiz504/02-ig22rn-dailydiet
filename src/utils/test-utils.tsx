import { FC, ReactNode } from 'react'
import { ThemeProvider } from 'styled-components/native'
import { render } from '@testing-library/react-native'

import { theme } from '~/styles'
import { NavigationContainer } from '@react-navigation/native'

const AllTheProviders: FC<{ children: ReactNode }> = ({ children }) => {
  return (
    <ThemeProvider theme={theme}>
      <NavigationContainer>{children}</NavigationContainer>
    </ThemeProvider>
  )
}

const ThemedWrapper: FC<{ children: ReactNode }> = ({ children }) => {
  return <ThemeProvider theme={theme}>{children}</ThemeProvider>
}

const customRender: typeof render = (ui, options) =>
  render(ui, { wrapper: AllTheProviders, ...options })

const renderThemed: typeof render = (ui, options) =>
  render(ui, { wrapper: ThemedWrapper, ...options })

const splitBorderColor = (borderColor: string) => ({
  borderTopColor: borderColor,
  borderRightColor: borderColor,
  borderBottomColor: borderColor,
  borderLeftColor: borderColor,
})

export * from '@testing-library/react-native'

export { customRender as render, renderThemed, splitBorderColor }
