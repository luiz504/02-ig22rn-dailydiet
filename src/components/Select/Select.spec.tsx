import { render, screen } from '~/utils/test-utils'
import { Select } from '.'

describe('Select Component', () => {
  describe('Label Component', () => {
    it('should render the Label Component', () => {
      const text = 'Some Text'
      render(<Select.Label testID="label">{text}</Select.Label>)

      const labelElement = screen.getByTestId('label')
      const { size, weight } = labelElement.props
      expect({ size, weight }).toMatchSnapshot('label')

      expect(screen.getByTestId('label')).toHaveTextContent(text)
    })
  })

  describe('Dot Component', () => {
    it('should render the Dot Component variant green', () => {
      render(<Select.Dot variant="green" testID="dot" />)

      expect(screen.getByTestId('dot').props.style).toMatchSnapshot('dot-green')
    })

    it('should render the Dot Component variant red', () => {
      render(<Select.Dot variant="red" testID="dot" />)

      expect(screen.getByTestId('dot').props.style).toMatchSnapshot('dot-red')
    })
  })

  describe('Button Component', () => {
    const getBtnProps = () =>
      screen.UNSAFE_getByProps({
        testID: 'button',
      }).props
    it('should render correctly the variant green', async () => {
      let selected = false

      render(
        <Select.Button isSelected={selected} variant="green" testID="button">
          <Select.Label>Text</Select.Label>
        </Select.Button>,
      )
      const btnElement = screen.getByTestId('button')

      const render1 = getBtnProps()

      expect({
        variant: render1.variant,
        isSelected: render1.isSelected,
      }).toEqual({
        variant: 'green',
        isSelected: false,
      })

      expect(btnElement.props.style).toMatchSnapshot('btn-green-not-selected')
      expect(
        screen.getByTestId('btn-content-wrapper').props.style,
      ).toMatchSnapshot('btn-content-wrapper')

      selected = true

      screen.rerender(
        <Select.Button isSelected={selected} variant="green" testID="button">
          <Select.Label>Text</Select.Label>
        </Select.Button>,
      )

      const render2 = getBtnProps()

      expect(render2.isSelected).toBe(true)
      expect(btnElement.props.style).toMatchSnapshot('btn-green-selected')
    })

    it('should render correctly the variant red', () => {
      let selected = false

      render(
        <Select.Button isSelected={selected} variant="red" testID="button">
          <Select.Label>Text</Select.Label>
        </Select.Button>,
      )

      const btnElement = screen.getByTestId('button')

      const render1 = getBtnProps()

      expect({
        variant: render1.variant,
        isSelected: render1.isSelected,
      }).toEqual({
        variant: 'red',
        isSelected: false,
      })

      expect(btnElement.props.style).toMatchSnapshot('btn-red-not-selected')

      selected = true

      screen.rerender(
        <Select.Button isSelected={selected} variant="red" testID="button">
          <Select.Label>Text</Select.Label>
        </Select.Button>,
      )

      const render2 = getBtnProps()

      expect(render2.isSelected).toBe(true)
      expect(btnElement.props.style).toMatchSnapshot('btn-red-selected')
    })
  })
})
