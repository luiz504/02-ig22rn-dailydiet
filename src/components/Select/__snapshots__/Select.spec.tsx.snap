// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Select Component Button Component should render correctly the variant green: btn-content-wrapper 1`] = `
{
  "alignItems": "center",
  "flexDirection": "row",
  "gap": 8,
  "justifyContent": "center",
}
`;

exports[`Select Component Button Component should render correctly the variant green: btn-green-not-selected 1`] = `
{
  "backgroundColor": "#EFF0F0",
  "borderBottomColor": "#EFF0F0",
  "borderBottomLeftRadius": 8,
  "borderBottomRightRadius": 8,
  "borderBottomWidth": 1.5,
  "borderLeftColor": "#EFF0F0",
  "borderLeftWidth": 1.5,
  "borderRightColor": "#EFF0F0",
  "borderRightWidth": 1.5,
  "borderTopColor": "#EFF0F0",
  "borderTopLeftRadius": 8,
  "borderTopRightRadius": 8,
  "borderTopWidth": 1.5,
  "overflow": "hidden",
  "paddingBottom": 16,
  "paddingLeft": 16,
  "paddingRight": 16,
  "paddingTop": 16,
}
`;

exports[`Select Component Button Component should render correctly the variant green: btn-green-selected 1`] = `
{
  "backgroundColor": "#E5F0DB",
  "borderBottomColor": "#639339",
  "borderBottomLeftRadius": 8,
  "borderBottomRightRadius": 8,
  "borderBottomWidth": 1.5,
  "borderLeftColor": "#639339",
  "borderLeftWidth": 1.5,
  "borderRightColor": "#639339",
  "borderRightWidth": 1.5,
  "borderTopColor": "#639339",
  "borderTopLeftRadius": 8,
  "borderTopRightRadius": 8,
  "borderTopWidth": 1.5,
  "overflow": "hidden",
  "paddingBottom": 16,
  "paddingLeft": 16,
  "paddingRight": 16,
  "paddingTop": 16,
}
`;

exports[`Select Component Button Component should render correctly the variant red: btn-red-not-selected 1`] = `
{
  "backgroundColor": "#EFF0F0",
  "borderBottomColor": "#EFF0F0",
  "borderBottomLeftRadius": 8,
  "borderBottomRightRadius": 8,
  "borderBottomWidth": 1.5,
  "borderLeftColor": "#EFF0F0",
  "borderLeftWidth": 1.5,
  "borderRightColor": "#EFF0F0",
  "borderRightWidth": 1.5,
  "borderTopColor": "#EFF0F0",
  "borderTopLeftRadius": 8,
  "borderTopRightRadius": 8,
  "borderTopWidth": 1.5,
  "overflow": "hidden",
  "paddingBottom": 16,
  "paddingLeft": 16,
  "paddingRight": 16,
  "paddingTop": 16,
}
`;

exports[`Select Component Button Component should render correctly the variant red: btn-red-selected 1`] = `
{
  "backgroundColor": "#F4E6E7",
  "borderBottomColor": "#BF3B44",
  "borderBottomLeftRadius": 8,
  "borderBottomRightRadius": 8,
  "borderBottomWidth": 1.5,
  "borderLeftColor": "#BF3B44",
  "borderLeftWidth": 1.5,
  "borderRightColor": "#BF3B44",
  "borderRightWidth": 1.5,
  "borderTopColor": "#BF3B44",
  "borderTopLeftRadius": 8,
  "borderTopRightRadius": 8,
  "borderTopWidth": 1.5,
  "overflow": "hidden",
  "paddingBottom": 16,
  "paddingLeft": 16,
  "paddingRight": 16,
  "paddingTop": 16,
}
`;

exports[`Select Component Dot Component should render the Dot Component variant green: dot-green 1`] = `
{
  "backgroundColor": "#639339",
  "borderBottomLeftRadius": 99999,
  "borderBottomRightRadius": 99999,
  "borderTopLeftRadius": 99999,
  "borderTopRightRadius": 99999,
  "height": 8,
  "width": 8,
}
`;

exports[`Select Component Dot Component should render the Dot Component variant red: dot-red 1`] = `
{
  "backgroundColor": "#BF3B44",
  "borderBottomLeftRadius": 99999,
  "borderBottomRightRadius": 99999,
  "borderTopLeftRadius": 99999,
  "borderTopRightRadius": 99999,
  "height": 8,
  "width": 8,
}
`;

exports[`Select Component Label Component should render the Label Component: label 1`] = `
{
  "size": "sm",
  "weight": "bold",
}
`;
